### PR TITLE
KAFKA-9206: throw a KafkaException when encountering CORRUPT_MESSAGE

### PR DESCRIPTION
**Pull request by agam**
- If the completed fetch has an error code signifying a _corrupt message_, throw a `KafkaException` that notes the _offset and the topic-partition_.
- Also added a test that triggers the warning and verifies it is thrown.

- Jira: https://issues.apache.org/jira/browse/KAFKA-9206

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

**Comment by agam** [comment]
Note: I found `LogCaptureAppender` in the `streams` module to be useful for verifying log messages, and _copied_ it to the `clients` module. Let me know if there's some common place I can place it instead.

File: build.gradle [hunk 1]
```diff
@@ -998,6 +998,7 @@ project(':clients') {
     testCompile libs.bcpkix
     testCompile libs.junit
     testCompile libs.mockitoCore
+    testCompile libs.log4j
```
Comment (ijuma) [inline build.gradle]: Why is this needed? Supposedly the `slf4jlog4j` dependency brings this too?

Comment (ijuma) [reply build.gradle]: I guess you're doing it because we refer to log4j at compile time now?

Comment (agam) [reply build.gradle]: yes, it's an import for `LogCaptureAppender`

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/Fetcher.java [hunk 2]
```diff
@@ -1272,6 +1272,8 @@ private CompletedFetch initializeCompletedFetch(CompletedFetch nextCompletedFetc
                 log.debug("Received unknown leader epoch error in fetch for partition {}", tp);
             } else if (error == Errors.UNKNOWN_SERVER_ERROR) {
                 log.warn("Unknown error fetching data for topic-partition {}", tp);
+            } else if (error == Errors.CORRUPT_MESSAGE) {
+                log.warn("Encountered corrupt message while fetching data for topic-partition {}", tp);
```
Comment (hachikuji) [inline clients/src/main/java/org/apache/kafka/clients/consumer/internals/Fetcher.java]: I think the main question here is whether we want to raise this error to the user. If we happen to see message corruption in the fetched data, we raise that to the user, so should this case be different?

Comment (agam) [reply clients/src/main/java/org/apache/kafka/clients/consumer/internals/Fetcher.java]: Ack. As discussed, will (1) change this to correspond to the current behavior for corruption in fetched data, i.e. throwing a `KafkaException` that wraps around a `CorruptRecordException`, and (2) remove the log message expectation.

**Comment by agam** [comment]
Updated, please take a look.

File: clients/src/test/java/org/apache/kafka/clients/consumer/internals/FetcherTest.java [hunk 3]
```diff
@@ -3835,6 +3835,31 @@ public void testFetchCompletedBeforeHandlerAdded() {
...
+                buildRecords(1L, 1, 1),
+                Errors.CORRUPT_MESSAGE,
+                100L, 0));
+        consumerClient.poll(time.timer(0));
+        assertTrue(fetcher.hasCompletedFetches());
+
+        // Trigger the exception.
+        assertThrows(KafkaException.class, () -> {
+            Map<TopicPartition, List<ConsumerRecord<byte[], byte[]>>> partitionRecords =
+                    fetchedRecords();
```
Comment (soondenana) [inline clients/src/test/java/org/apache/kafka/clients/consumer/internals/FetcherTest.java]: nit: just calling fetchedRecords() should work here, or is compiler not liking it and forcing the return value assignment?

**Review by soondenana — COMMENTED** [review]
> i.e. throwing a KafkaException that wraps around a CorruptRecordException

Are you planing to have a CorruptRecordException exception too?

File: clients/src/test/java/org/apache/kafka/clients/consumer/internals/FetcherTest.java (hunk 3 above)
Comment (agam) [reply clients/src/test/java/org/apache/kafka/clients/consumer/internals/FetcherTest.java]: Agree, made the change to discard the return value.

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/Fetcher.java (hunk 2 above)
Comment (agam) [reply clients/src/main/java/org/apache/kafka/clients/consumer/internals/Fetcher.java]: Correction: _just_ a `KafkaException` (the wrapping around I referred to was the _text_ of the `CorruptRecordException`, not the exception itself)

**Comment by agam** [comment]
> > i.e. throwing a KafkaException that wraps around a CorruptRecordException
> 
> Are you planing to have a CorruptRecordException exception too?

No, clarified with a comment above too, I initially thought we were wrapping the exception in the "data corruption" case, but we're just wrapping its _text_.

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/Fetcher.java [hunk 4]
```diff
@@ -1272,6 +1272,11 @@ private CompletedFetch initializeCompletedFetch(CompletedFetch nextCompletedFetc
                 log.debug("Received unknown leader epoch error in fetch for partition {}", tp);
             } else if (error == Errors.UNKNOWN_SERVER_ERROR) {
                 log.warn("Unknown error fetching data for topic-partition {}", tp);
+            } else if (error == Errors.CORRUPT_MESSAGE) {
+                throw new KafkaException("Encountered corrupt message when fetching offset "
```
Comment (hachikuji) [inline clients/src/main/java/org/apache/kafka/clients/consumer/internals/Fetcher.java:1277]: While we're at it, could we add the fetch offset to the IllegalStateException message below and the `UNKNOWN_SERVER_ERROR` log message above?

Comment (agam) [reply clients/src/main/java/org/apache/kafka/clients/consumer/internals/Fetcher.java:1277]: Done.

**Comment by hachikuji** [comment]
ok to test

**Comment by agam** [comment]
Test failure seems unrelated to this change: `org.apache.kafka.clients.admin.KafkaAdminClientTest.testDefaultApiTimeoutOverride` ?

(will rerun tests once)

**Comment by agam** [comment]
ok to test

**Comment by agam** [comment]
retest this please

**Comment by hachikuji** [comment]
The test failure is known flaky. I have a fix in #8101.

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/Fetcher.java [hunk 5]
```diff
@@ -1271,9 +1271,20 @@ private CompletedFetch initializeCompletedFetch(CompletedFetch nextCompletedFetc
             } else if (error == Errors.UNKNOWN_LEADER_EPOCH) {
                 log.debug("Received unknown leader epoch error in fetch for partition {}", tp);
             } else if (error == Errors.UNKNOWN_SERVER_ERROR) {
-                log.warn("Unknown error fetching data for topic-partition {}", tp);
+                log.warn("Unknown error fetching data while fetching offset "
```
Comment (hachikuji) [inline clients/src/main/java/org/apache/kafka/clients/consumer/internals/Fetcher.java]: nit: for log messages, can we use the `{}` placeholders?

**Comment by hachikuji** [comment]
retest this please

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/Fetcher.java (hunk 5 above)
Comment (agam) [reply clients/src/main/java/org/apache/kafka/clients/consumer/internals/Fetcher.java]: Done

**Comment by hachikuji** [comment]
ok to test

**Comment by agam** [comment]
Couple of tests that failed in this run but not the previous run, likely flaky: 

- `kafka.admin.DescribeConsumerGroupTest.testDescribeGroupMembersWithShortInitializationTimeout`
- `kafka.admin.DescribeConsumerGroupTest.testDescribeGroupWithShortInitializationTimeout`

**Review by hachikuji — APPROVED** [review]
LGTM. Thanks for the PR!

**Comment by hachikuji** [comment]
@agam One other small thing before we merge. Can you add CORRUPT_MESSAGE to the list of error codes in the javadoc for `FetchResponse`?

**Comment by hachikuji** [comment]
retest this please

**Comment by hachikuji** [comment]
ok to test

**Comment by hachikuji** [comment]
retest this please

**Comment by agam** [comment]
Two test failures:

- `kafka.admin.DescribeConsumerGroupTest.testDescribeGroupMembersWithShortInitializationTimeout` (same as earlier)
- `kafka.api.ConsumerBounceTest.testRollingBrokerRestartsWithSmallerMaxGroupSizeConfigDisruptsBigGroup` (newer failure in this run)
